raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientPreVoteTest.java

```diff
@@ -0,0 +1,543 @@
+        context.assertSentVoteResponse(Errors.NONE, epoch, OptionalInt.empty(), true);
+        context.assertVotedCandidate(epoch, votedCandidateKey.id());
+    }
+
+    @Test
+    public void testHandlePreVoteRequestAsCandidate() throws Exception {
+        int localId = randomReplicaId();
+        ReplicaKey otherNodeKey = replicaKey(localId + 1, true);
+        int leaderEpoch = 2;
+        Set<Integer> voters = Set.of(localId, otherNodeKey.id());
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
+            .withVotedCandidate(leaderEpoch, ReplicaKey.of(localId, ReplicaKey.NO_DIRECTORY_ID))
+            .withKip853Rpc(true)
+            .build();
+
+        context.pollUntilRequest();
+
+        // candidate should grant pre-vote requests with the same epoch if log is up-to-date
+        context.deliverRequest(context.preVoteRequest(leaderEpoch, otherNodeKey, leaderEpoch, 1));
+        context.pollUntilResponse();
+
+        context.assertSentVoteResponse(Errors.NONE, leaderEpoch, OptionalInt.empty(), true);
+        context.assertVotedCandidate(leaderEpoch, localId);
+
+        // candidate will transition to unattached if pre-vote request has a higher epoch
+        context.deliverRequest(context.preVoteRequest(leaderEpoch + 1, otherNodeKey, leaderEpoch + 1, 1));
+        context.pollUntilResponse();
+
+        context.assertSentVoteResponse(Errors.NONE, leaderEpoch + 1, OptionalInt.of(-1), true);
+        assertTrue(context.client.quorum().isUnattached());
+    }
+
+    @Test
+    public void testHandlePreVoteRequestAsUnattachedObserver() throws Exception {
+        int localId = randomReplicaId();
+        int epoch = 2;
+        ReplicaKey replica1 = replicaKey(localId + 1, true);
+        ReplicaKey replica2 = replicaKey(localId + 2, true);
+        Set<Integer> voters = Set.of(replica1.id(), replica2.id());
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
+            .withUnknownLeader(epoch)
+            .withKip853Rpc(true)
+            .build();
+
+        context.deliverRequest(context.preVoteRequest(epoch, replica1, epoch, 1));
+        context.pollUntilResponse();
+
+        assertTrue(context.client.quorum().isUnattached());
+        context.assertSentVoteResponse(Errors.NONE, epoch, OptionalInt.empty(), true);
+
+        // if same replica sends another pre-vote request for the same epoch, it should be granted
+        context.deliverRequest(context.preVoteRequest(epoch, replica1, epoch, 1));
+        context.pollUntilResponse();
+
+        assertTrue(context.client.quorum().isUnattached());
+        context.assertSentVoteResponse(Errors.NONE, epoch, OptionalInt.empty(), true);
+
+        // if different replica sends a pre-vote request for the same epoch, it should be granted
+        context.deliverRequest(context.preVoteRequest(epoch, replica2, epoch, 1));
+        context.pollUntilResponse();
```

Comment:
We should also test that a replica outside of the voter set can send a pre-vote request and it will be granted accordingly.

This comment applies to a few of the other tests.